cpp/include/cudf/detail/nvtx/ranges.hpp

```diff
@@ -59,3 +74,52 @@ using scoped_range = ::nvtx3::scoped_range_in<libcudf_domain>;
  * ```
  */
 #define CUDF_FUNC_RANGE() NVTX3_FUNC_RANGE_IN(cudf::libcudf_domain)
+
+/**
+ * @brief Convenience macro for generating an NVTX scoped range in the `libcudf` domain with stream
+ * synchronization at the end of the scope.
+ *
+ * This is an internal development utility, which should be compiled to nothing in production
+ * builds. The NVTX range is generated and stream is synchronized only if the macro
+ * `CUDF_ENABLE_NVTX_DEBUG_RANGE` is explicitly defined.
+ *
+ * Example:
+ * ```
+ * {
+ *    CUDF_SCOPED_RANGE_DEBUG("Search keys", stream);
+ *    ...
+ * }
+ * ```
+ */
+#ifdef CUDF_ENABLE_NVTX_DEBUG_RANGE
+#define CUDF_SCOPED_RANGE_DEBUG(str, stream) \
+  [[maybe_unused]] cudf::scoped_range_sync __range_sync{str, stream};
+#else  // no-op
+#define CUDF_SCOPED_RANGE_DEBUG(str, stream)
+#endif
+
+/**
+ * @brief Convenience macro for generating an NVTX range in the `libcudf` domain from the lifetime
+ * of a function, synchronizing the given stream at the end of the function.
+ *
+ * This is an internal development utility, which should be compiled to nothing in production
+ * builds. The NVTX range is generated and stream is synchronized only if the macro
+ * `CUDF_ENABLE_NVTX_DEBUG_RANGE` is explicitly defined.
+ *
+ * Similar to `CUDF_FUNC_RANGE`, this uses the name of the immediately enclosing function returned
+ * by `__func__` to name the range.
+ *
+ * Example:
+ * ```
+ * void some_function(rmm::cuda_stream_view stream){
+ *    CUDF_FUNC_RANGE_DEBUG(stream);
+ *    ...
+ * }
+ * ```
+ */
+#ifdef CUDF_ENABLE_NVTX_DEBUG_RANGE
+#define CUDF_FUNC_RANGE_DEBUG(stream) \
```

Comment:
I feel like `CUDF_FUNC_RANGE_DEBUG()` should simply translate to `CUDF_FUNC_RANGE()` in debug mode instead of syncing a stream. 

Perhaps we could move this to a new macro called `CUDF_FUNC_RANGE_SYNC_DEBUG(stream)`